### PR TITLE
stubs and test for rnaseq modules - part V

### PR DESCRIPTION
**Pull request by maxulysse**
Improve/fix stub, stub tests + single snapshot + snapshot versions:

- `fastqc`
- `qualimap/bamqc`
- `qualimap/bamqccram`
- `qualimap/rnaseq`
- `picard/markduplicates`
- `samtools/flagstat`
- `samtools/idxstats`
- `samtools/index`
- `samtools/sort`
- `samtools/stats`
- `star/align`
- `star/genomegenerate`
- `stringtie/stringtie`
- `ucsc/bedgraphtobigwig`

Updating snapshot for depending subworkflows:
- `bam_dedup_stats_samtools_umitools`
- `bam_markduplicates_picard`
- `bam_sort_stats_samtools`
- `fastq_fastqc_umitools_fastp`
- `fastq_fastqc_umitools_trimgalore`

## PR checklist

Closes #XXX 

- [ ] This comment contains a description of changes (with reason).
- [ ] If you've fixed a bug or added code that should be tested, add tests!
- [ ] If you've added a new tool - have you followed the module conventions in the [contribution docs](https://github.com/nf-core/modules/tree/master/.github/CONTRIBUTING.md)
- [ ] If necessary, include test data in your PR.
- [ ] Remove all TODO statements.
- [ ] Emit the `versions.yml` file.
- [ ] Follow the naming conventions.
- [ ] Follow the parameters requirements.
- [ ] Follow the input/output options guidelines.
- [ ] Add a resource `label`
- [ ] Use BioConda and BioContainers if possible to fulfil software requirements.
- Ensure that the test works with either Docker / Singularity. Conda CI tests can be quite flaky:
  - For modules:
    - [ ] `nf-core modules test <MODULE> --profile docker`
    - [ ] `nf-core modules test <MODULE> --profile singularity`
    - [ ] `nf-core modules test <MODULE> --profile conda`
  - For subworkflows:
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile docker`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile singularity`
    - [ ] `nf-core subworkflows test <SUBWORKFLOW> --profile conda`

### CODE REVIEW

File: modules/nf-core/qualimap/bamqc/tests/main.nf.test [hunk 1]
```diff
@@ -9,11 +9,35 @@ nextflow_process {
...
+                """
             }
+        }
+
+        then {
+            assertAll (
+                { assert process.success },
+                { assert snapshot(
+                    file("${process.out.results[0][1]}/qualimapReport.html").name,
+                    path("${process.out.results[0][1]}/genome_results.txt"),
```
Comment (SPPearce) [inline modules/nf-core/qualimap/bamqc/tests/main.nf.test:31]: why is one of these file, the other path?

**Review by SPPearce — COMMENTED** [review]
This looks like it is just changes to the tests and a few more stub tests, no actual changes to the stub generation?

File: modules/nf-core/qualimap/bamqc/tests/main.nf.test (hunk 1 above)
Comment (maxulysse) [reply modules/nf-core/qualimap/bamqc/tests/main.nf.test:31]: The first one is just snapshoting the file name, the second the file content

**Comment by maxulysse** [comment]
> This looks like it is just changes to the tests and a few more stub tests, no actual changes to the stub generation?

Yes, I think I did update the actual stub generations in previous PRs

**Review by SPPearce — APPROVED** [review]

File: modules/nf-core/qualimap/bamqccram/main.nf [hunk 2]
```diff
@@ -57,4 +57,15 @@ process QUALIMAP_BAMQCCRAM {
...
     """
+
+    stub:
+    prefix = task.ext.suffix ? "${meta.id}${task.ext.suffix}" : "${meta.id}"
+    """
+    mkdir ${prefix}
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        qualimap: \$(echo \$(qualimap 2>&1) | sed 's/^.*QualiMap v.//; s/Built.*\$//')
```
Comment (SPPearce) [inline modules/nf-core/qualimap/bamqccram/main.nf:68]: This should have samtools too.

File: modules/nf-core/qualimap/bamqccram/main.nf [hunk 3]
```diff
@@ -57,4 +57,15 @@ process QUALIMAP_BAMQCCRAM {
         samtools: \$(echo \$(samtools --version 2>&1) | sed 's/^.*samtools //; s/Using.*\$//')
     END_VERSIONS
     """
+
+    stub:
+    prefix = task.ext.suffix ? "${meta.id}${task.ext.suffix}" : "${meta.id}"
```
Comment (SPPearce) [inline modules/nf-core/qualimap/bamqccram/main.nf:62]: Is there no separator here? And why does it need suffix?

File: modules/nf-core/star/genomegenerate/tests/main.nf.test [hunk 4]
```diff
@@ -45,24 +43,23 @@ nextflow_process {
...
             }
         }
 
         then {
             assertAll(
                 { assert process.success },
-                { assert snapshot(file(process.out.index[0][1]).listFiles().collect { it.getName() }.sort().toString()).match("fasta_gtf_stub_index") },
-                { assert snapshot(process.out.versions).match("fasta_gtf_stub_versions") }
+                { assert snapshot(file(process.out.index[0][1]).listFiles().collect { it.getName() }.sort().toString()).match("fasta_index") },
+                { assert snapshot(process.out.versions).match("fasta_versions") }
```
Comment (SPPearce) [inline modules/nf-core/star/genomegenerate/tests/main.nf.test:55]: Can you put these into the same snapshot?

File: modules/nf-core/star/genomegenerate/tests/main.nf.test [hunk 5]
```diff
@@ -45,24 +43,23 @@ nextflow_process {
...
             )
         }
     }
 
-    test("fasta") {
+    test("fasta_gtf_stub") {
+
```
Comment (SPPearce) [inline modules/nf-core/star/genomegenerate/tests/main.nf.test:61]:
```suggestion
```

File: modules/nf-core/qualimap/bamqccram/main.nf (hunk 3 above)
Comment (maxulysse) [reply modules/nf-core/qualimap/bamqccram/main.nf:62]: what about the no separator?

For the second question, I just copied the line setting up the prefix.
I didn't looked about why it's doing that.

File: modules/nf-core/qualimap/bamqccram/main.nf (hunk 2 above)
Comment (maxulysse) [reply modules/nf-core/qualimap/bamqccram/main.nf:68]: good point, I'll update that

File: modules/nf-core/qualimap/bamqccram/main.nf (hunk 3 above)
Comment (maxulysse) [reply modules/nf-core/qualimap/bamqccram/main.nf:62]: scratch that, I have no idea where this is coming from, bad copy paste I'd say, I'll fix it right away

**Review by SPPearce — APPROVED** [review]

File: modules/nf-core/samtools/sort/main.nf
[the remark this answers is not in the document]
Comment (maxulysse) [reply modules/nf-core/samtools/sort/main.nf:1]: That's not implemented in the current module, and I don't think we should allow for people to work on sam files anyway

**Pull request merged**